lib/rouster.rb

```diff
@@ -292,60 +293,87 @@ def inspect
-
-    cmd = sprintf('%s%s; echo ec[$?]', self.uses_sudo? ? 'sudo ' : '', command)
-    @logger.info(sprintf('vm running: [%s]', cmd)) # TODO decide whether this should be changed in light of passthroughs.. 'remotely'?
+    @logger.info( sprintf( 'vm running: [%s]', cmd[:final_command] ) ) # TODO decide whether this should be changed in light of passthroughs.. 'remotely'?
 
     0.upto(@retries) do |try|
       begin
         if self.is_passthrough? and self.passthrough[:type].eql?(:local)
-          output = `#{cmd}`
+          cmd[:stdout]   = `#{cmd[:final_command]}`
+          cmd[:exitcode] = $?
         else
-          output = @ssh.exec!(cmd)
+          cmd = remote_exec( cmd )
         end
-
         break
       rescue => e
-        @logger.error(sprintf('failed to run [%s] with [%s], attempt[%s/%s]', cmd, e, try, retries)) if self.retries > 0
+        @logger.error(sprintf('failed to run [%s] with [%s], attempt[%s/%s]', cmd[:final_command], e, try, retries)) if self.retries > 0
         sleep 10 # TODO need to expose this as a variable
       end
-
     end
 
-    if output.nil?
-      output    = "error gathering output, last logged output[#{self.get_output()}]"
-      @exitcode = 256
-    elsif output.match(/ec\[(\d+)\]/)
-      @exitcode = $1.to_i
-      output.gsub!(/ec\[(\d+)\]\n/, '')
-    else
-      @exitcode = 1
+    if cmd[:stdout].nil?
+      cmd[:stdout]   = "error gathering output, last logged output[#{self.get_ssh_stdout()}]"
+      cmd[:exitcode] = 256
+    elsif cmd[:exitcode].nil?
+      cmd[:exitcode] = 255
     end
 
-    self.output.push(output)
-    @logger.debug(sprintf('output: [%s]', output))
+    self.ssh_stdout.push(   cmd[:stdout]   )
+    self.ssh_stderr.push(   cmd[:stderr]   )
+    self.ssh_exitcode.push( cmd[:exitcode] )
+    @logger.debug( sprintf( 'output: [%s]',     cmd[:stdout] ) )
+    @logger.debug( sprintf( 'ssh_stderr: [%s]', cmd[:stderr] ) )
 
-    unless expected_exitcode.member?(@exitcode)
+    unless cmd[:expected_exitcode].member?( cmd[:exitcode] )
       # TODO technically this could be a 'LocalPassthroughExecutionError' now too if local passthrough.. should we update?
-      raise RemoteExecutionError.new("output[#{output}], exitcode[#{@exitcode}], expected[#{expected_exitcode}]")
+      raise RemoteExecutionError.new("output[#{cmd[:stdout]}], exitcode[#{cmd[:exitcode]}], expected[#{cmd[:expected_exitcode]}]")
     end
 
-    @exitcode ||= 0
-    output
+    cmd[:stdout]
+  end
+
+  def remote_exec( cmd )
+    @ssh.open_channel do |channel|
+      channel.exec( cmd[:final_command] ) do |ch, success|
+        unless success
+          error = "FAILED: couldn't execute command remotely [#{cmd[:final_command]}]"
+          @logger.error( error )
+          raise RemoteExecutionError.new( error )
+        end
+        channel.on_data do |ch, data|
+          cmd[:stdout] << data
+        end
+        channel.on_extended_data do |ch, type, data|
```

Comment:
can you change `ch` to `_ch` and `type` to `_type` here to make it clear that we're not using these parameters?

also, how do you feel about rewriting these 3 blocks as:

```rb

channel.on_data { |_ch, data| cmd[:stdout] << data }
channel.on_extended_data { |_ch, _type, data| cmd[:stderr] << data }
channel.on_request('exit-status') do { |_ch, data| cmd[:exitcode] = data.read_long } 
```

i'm not sure i like it better based on indentation, but single line do blocks are something i've been moving away from